src/python/pants/build_graph/address.py

```diff
@@ -233,10 +227,31 @@ def __init__(
         """
         self.spec_path = spec_path
         self._relative_file_path = relative_file_path
+
         # If the target_name is the same as the default name would be, we normalize to None.
-        self._target_name = (
-            target_name if target_name and target_name != os.path.basename(self.spec_path) else None
-        )
+        self._target_name: Optional[str]
+        if target_name and target_name != os.path.basename(self.spec_path):
+            banned_chars = BANNED_CHARS_IN_TARGET_NAME & set(target_name)
+            deprecated_banned_chars = banned_chars & set(r"/\:")
+            if deprecated_banned_chars:
+                warn_or_error(
+                    removal_version="2.2.0.dev1",
```

Comment:
Don't we need to ban `/` in 2.0.1 for Pants to work properly? How about banning that outright, and deprecating the others for 2.2.0. We'll backport this PR into 2.0.1.

Reply:
Even in @rcuza's case (#11106), things mostly worked, so I wasn't sure that turning this into a hard error was strictly necessary here. This will hopefully also make it easier to cherry-pick ~verbatim.

Reply:
Hm, I thought Raúl ended up needing to rename the targets to get it working? I recommend that we fail eagerly on `/` in 2.x.x, but totally fine to deprecate `\` and `:`.

Reply:
I only hit the issue when I used the `git` features (`changed` something or other).